File: C++/Array/Smallest Subset with Greater Sum.cpp
// problem Link: https://practice.geeksforgeeks.org/problems/smallest-subset-with-greater-sum/1

/*
Approach: 
Get the sum of whole vector then 
Sorts the vector in descending order then
traverse the vector and check 
if currentSum < restElementSum then
	currentSum+=Arr[i] 		//currSum Stores the sum of subset
	restElementSum -= Arr[i]		//restElement Stores the sum of other elements
	count++ 				//size of current subset
return count ;
*/

/* 
Time Complexity: O(N log N)
Space Complexity: O(1)
*/

class Solution{
    public:
    int minSubset(vector<int> &Arr, int N){
        sort(Arr.begin(),Arr.end(),greater<int>());
        long long sum =0;
        int count = 0;
        long long MaxSum = 0;
        for(int i =0;i<N;i++){
            sum +=Arr[i];
        }
        for(int i=0;i<N;i++){
            if(MaxSum<=sum){
                MaxSum+=Arr[i];
                sum-=Arr[i];
                count++;
            }
            else{
	break;
            }
        }
        return count;
    }
};

File: Java/Array/Smallest Subset with Greater Sum.java
// problem Link: https://practice.geeksforgeeks.org/problems/smallest-subset-with-greater-sum/1

/***
 * Approach: we need to find minimum number of elements from array
 * such that their sum is greater sum of all other elements from array.
 * 
 * STEP 1: to find minimum number of elements, we need to get maximum elements,
 * so I sort the array
 * 
 * STEP 2: then take three variables-sum, temp1 and count
 * temp1 stores sum of largest elements of array
 * sum stores sum(all other elements of array)
 * count store number of elements that fulfilled this coniditon.
 */


/****
 * Time Complexity: O(N log N)
 * Space Complexity: O(1)
 */




import java.util.Arrays;

class Solution { 
    int minSubset(int[] Arr,int N) { 
        Arrays.sort(Arr);
        long sum=0;
        for (int i=0;i<N;i++){
            sum+=Arr[i];
        }
        sum=sum-Arr[N-1];
        long temp1=Arr[N-1];
        int count=1;
        int i=N-2;
        while(i>=0 && temp1<=sum)
        {
            temp1+=Arr[i];
            sum-=Arr[i];
            count++;
            i--;
        }
        return count;
    }
}